Checked for bad url in manually added forms and removed, and notified user

pratikmjoshi opened the pull request:
Fixes #709 
When filling a blank form, when forms are scanned, the prescence of bad url throws illegalargumentexception and the user is notified that it has a bad url.

@yanokwa Could you please verify this pr?

pratikmjoshi committed:
Merge pull request #1 from opendatakit/master

Update from original

pratikmjoshi committed:
Merge remote-tracking branch 'upstream/master'

pratikmjoshi committed:
Merge remote-tracking branch 'upstream/master'

pratikmjoshi committed:
Merge branch 'master' of https://github.com/opendatakit/collect

pratikmjoshi committed:
Merge branch 'master' of https://github.com/opendatakit/collect

pratikmjoshi committed:
Merge branch 'master' of https://github.com/opendatakit/collect

pratikmjoshi committed:
Merge branch 'master' of https://github.com/opendatakit/collect

pratikmjoshi committed:
Removed bad url forms from blank forms and notified user of bad url form

yanokwa commented on collect_app/src/main/java/org/odk/collect/android/tasks/DiskSyncTask.java:
```diff
@@ -301,7 +305,11 @@ public ContentValues buildContentValues(File formDefFile) throws IllegalArgument
+            if(checkValidSubmissionURI(submission)) {
+                updateValues.put(FormsColumns.SUBMISSION_URI, submission);
+            } else {
+                throw new IllegalArgumentException(formDefFile.getName() + " :: bad url");
```
This message should be put into strings.xml so it can be localized. Might be easier to just refer to `url_error`.

lognaturel replied:
Note that this is not required for all exception text. It's just that this particular exception text gets shown in the UI!

yanokwa commented on collect_app/src/main/java/org/odk/collect/android/tasks/DiskSyncTask.java:
```diff
@@ -326,4 +334,7 @@ protected void onPostExecute(String result) {
         }
     }
 
+    public boolean checkValidSubmissionURI(String submissionURI) {
```
Use UrlUtils.isValidUrl so we have one implementation.

lognaturel replied:
I think what @yanokwa is suggesting is that instead of creating this new method, it would be preferable to just use `UrlUtils.isValidUrl` in the code. It turns out that both do about the same check. Both allow a lot that doesn't actually end up working in this case so we should investigate whether we can improve on that implementation.

yanokwa commented:
This does a good job of catching if a bad form is put on the SD card, but it doesn't catch the problem if the bad form is downloaded from a server. I'm not sure where the best place in the code to do this. `DownloadFormsTask.java` might be a good place to start.

pratikmjoshi committed:
Simplified url verify method and set error message to url_error

pratikmjoshi commented:
@yanokwa @lognaturel Sorry for the delay.I've made the changes requested. 
Also, if I use UrlUtils.isValidUrl, then the bad form from server is not removed. However, when I used Patterns.WEB_URL.matcher() function, bad forms from server are also removed once downloaded. So I've removed the method, but I've not used UrlUtils.
I placed the bug.xml file in my google drive and downloaded it. It got removed as soon as I clicked "Fill Blank Form" and I was notified. I don't think there's a need to make any changes to DownloadFormsTask, because DiscSyncTask runs for all unparsed files, and any form downloaded from the server will get removed if it has a bad url.

pratikmjoshi committed:
Removed forms with bad url and notified user

pratikmjoshi committed:
Removed unused import UrlUtils

yanokwa commented:
@pratikmjoshi I think we are making good progress! 

I have added a test form on https://nafundi-test.appspot.com. This form has a bad URL (e.g., `not-a-url`). Once this form is downloaded, I can fill it out, submit it, and it'll fail submission with Client Protocol Exception. This happens with your PR.

pratikmjoshi commented:
@yanokwa Yes I tried it out and got the same problem. When I try to access the odk/forms directory(on my phone) through my laptop, the form xmls are not visible. However, when I view the directory through the file manager on my phone, they are visible. Why is this so?

yanokwa commented:
@pratikmjoshi The computer-android file browsing tools are terrible. I'd recommend you use `adb shell` and navigate that way.

lognaturel commented:
@pratikmjoshi @yanokwa I'm not clear on the status of this PR! Please update me, @yanokwa are you asking for a change?

pratikmjoshi commented:
@lognaturel The current PR solves the problem of removing manually added bad forms, as well as bad forms downloaded from google drive. However, it doesn't remove bad forms downloaded from the server.

lognaturel commented:
@yanokwa Can you please clarify whether this is sufficient for this pull request or whether behavior needs to be added?

yanokwa committed:
Merge remote-tracking branch 'upstream/master' into badurl

yanokwa committed:
Consistent url validation, reuse string, clean imports

yanokwa commented:
@pratikmjoshi I'm very sorry I wasn't able to get to this sooner. And because I let this sit so long, I had to do something I rarely do, I pushed the changes I'd like to see into your branch so we can merge it. The changes are:
* Using UrlUtils.isValidUrl() so we have consistent validation
* Removed unnessary logging and imports
* Reused xform_parse_error string

yanokwa commented:
As to the issue of the bad forms not working in Aggregate, I've filed that separately at https://github.com/opendatakit/collect/issues/873. Perhaps you can work on it there? I'll do my best to get that code reviewed quickly. ✨Thanks again for your contribution!✨
